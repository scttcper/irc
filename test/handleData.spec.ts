import { describe, it, expect, jest } from '@jest/globals';

import { setupMockClient } from './helpers';

const messages = `:origin.irc.xx-net.org NOTICE * :*** Looking up your hostname...
:origin.irc.xx-net.org NOTICE * :*** Found your hostname (cached)
PING :EAA41EAE
|chunk|:origin.irc.xx-net.org 001 testbot :Welcome to the xx-Net IRC Network testbot!nodebot@nothing.sanic.net
:origin.irc.xx-net.org 002 testbot :Your host is origin.irc.xx-net.org, running version UnrealIRCd-5.0.4
:origin.irc.xx-net.org 003 testbot :This server was created Fri Apr 24 2020 at 13:04:49 UTC
:origin.irc.xx-net.org 004 testbot origin.irc.xx-net.org UnrealIRCd-5.0.4 iowrsxzdHtIDZRqpWGTSB lvhopsmntikraqbeIHzMQNRTOV|chunk|KDdGLPZSCcf
:origin.irc.xx-net.org 005 testbot AWAYLEN=307 CASEMAPPING=ascii CHANLIMIT=#:30 CHANMODES=beI,kLf,lH,psmntirzMQNRTOVKDdGPZSCc CHANNELLEN=32 CHANTYPES=# DEAF=d ELIST=MNUCT EXCEPTS EXTBAN=~,ptmTSOcarnqjf HCN INVEX :are supported by this server
:origin.irc.xx-net.org 005 testbot KICKLEN=307 KNOCK MAP MAXCHANNELS=30 MAXLIST=b:60,e:60,I:60 MAXNICKLEN=30 MINNICKLEN=0 MODES=12 NAMESX NETWORK=xx-Net NICKLEN=30 PREFIX=(qaohv)~&@%+ :are supported by this server
:origin.irc.co|chunk|rrupt-net.org 005 testbot QUITLEN=307 SAFELIST SILENCE=15 STATUSMSG=~&@%+ TARGMAX=DCCALLOW:,ISON:,JOIN:,KICK:4,KILL:,LIST:,NAMES:1,NOTICE:1,PART:,PRIVMSG:4,SAJOIN:,SAPART:,USERHOST:,USERIP:,WATCH:,WHOIS:1,WHOWAS:1 TOPICLEN=360 UHNAMES USERIP WALLCHOPS WATCH=128 WATCHOPTS=A WHOX :are supported by this server
:origin.irc.xx-net.org 396 testbot xx-xx.nothing.sanic.net :is now your displayed host
:origin.irc.xx-net.org NOTICE testbot :*** You are connected to origi|chunk|n.irc.xx-net.org with TLSv1.3-TLS_CHACHA20_POLY1305_SHA256
:origin.irc.xx-net.org 251 testbot :There are 1 users and 530 invisible on 6 servers
:origin.irc.xx-net.org 252 testbot 40 :operator(s) online
:origin.irc.xx-net.org 253 testbot 1 :unknown connection(s)
:origin.irc.xx-net.org 254 testbot 87 :channels formed
:origin.irc.xx-net.org 255 testbot :I have 170 clients and 1 servers
:origin.irc.xx-net.org 265 testbot 170 411 :Curre|chunk|nt local users 170, max 411
:origin.irc.xx-net.org 266 testbot 531 756 :Current global users 531, max 756
:origin.irc.xx-net.org 422 testbot :MOTD File is missing
:testbot MODE testbot :+iwxz
|chunk|:NickServ!nickserv@nickserv.services.irc.xx-net.org NOTICE testbot :This nickname is registered and protected.  If it is your
:NickServ!nickserv@nickserv.services.irc.xx-net.org NOTICE testbot :nick, type /msg NickServ IDENTIFY password.  Otherwise,
:NickServ!nickserv@nickserv.services.irc.xx-net.org NOTICE testbot :please choose a different nick.
:NickServ!nickserv@nickserv.services.irc.xx-net.org NOTICE testbot :If you do not change within 1 minute, I wil|chunk|l change your nick.
|chunk|`;

describe('handle data', () => {
  it('should handle initial connection', () => {
    const client = setupMockClient('testbot');
    const emitSpy = jest.spyOn(client, 'emit');
    for (const chunk of messages.split('|chunk|')) {
      client.handleData(chunk);
    }

    expect(emitSpy.mock.calls).toMatchSnapshot();
  });
});
