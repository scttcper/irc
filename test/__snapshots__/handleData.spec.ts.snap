// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handle data should handle initial connection 1`] = `
Array [
  Array [
    "raw",
    Object {
      "args": Array [
        "*",
        "*** Looking up your hostname...",
      ],
      "command": "NOTICE",
      "commandType": "normal",
      "prefix": "origin.irc.xx-net.org",
      "rawCommand": "NOTICE",
      "server": "origin.irc.xx-net.org",
    },
  ],
  Array [
    "notice",
    undefined,
    "*",
    "*** Looking up your hostname...",
    Object {
      "args": Array [
        "*",
        "*** Looking up your hostname...",
      ],
      "command": "NOTICE",
      "commandType": "normal",
      "prefix": "origin.irc.xx-net.org",
      "rawCommand": "NOTICE",
      "server": "origin.irc.xx-net.org",
    },
  ],
  Array [
    "raw",
    Object {
      "args": Array [
        "*",
        "*** Found your hostname (cached)",
      ],
      "command": "NOTICE",
      "commandType": "normal",
      "prefix": "origin.irc.xx-net.org",
      "rawCommand": "NOTICE",
      "server": "origin.irc.xx-net.org",
    },
  ],
  Array [
    "notice",
    undefined,
    "*",
    "*** Found your hostname (cached)",
    Object {
      "args": Array [
        "*",
        "*** Found your hostname (cached)",
      ],
      "command": "NOTICE",
      "commandType": "normal",
      "prefix": "origin.irc.xx-net.org",
      "rawCommand": "NOTICE",
      "server": "origin.irc.xx-net.org",
    },
  ],
  Array [
    "raw",
    Object {
      "args": Array [
        "EAA41EAE",
      ],
      "command": "PING",
      "commandType": "normal",
      "rawCommand": "PING",
    },
  ],
  Array [
    "ping",
    "EAA41EAE",
  ],
  Array [
    "raw",
    Object {
      "args": Array [
        "testbot",
        "Welcome to the xx-Net IRC Network testbot!nodebot@nothing.sanic.net",
      ],
      "command": "rpl_welcome",
      "commandType": "reply",
      "prefix": "origin.irc.xx-net.org",
      "rawCommand": "001",
      "server": "origin.irc.xx-net.org",
    },
  ],
  Array [
    "registered",
    Object {
      "args": Array [
        "testbot",
        "Welcome to the xx-Net IRC Network testbot!nodebot@nothing.sanic.net",
      ],
      "command": "rpl_welcome",
      "commandType": "reply",
      "prefix": "origin.irc.xx-net.org",
      "rawCommand": "001",
      "server": "origin.irc.xx-net.org",
    },
  ],
  Array [
    "raw",
    Object {
      "args": Array [
        "testbot",
        "Your host is origin.irc.xx-net.org, running version UnrealIRCd-5.0.4",
      ],
      "command": "rpl_yourhost",
      "commandType": "reply",
      "prefix": "origin.irc.xx-net.org",
      "rawCommand": "002",
      "server": "origin.irc.xx-net.org",
    },
  ],
  Array [
    "raw",
    Object {
      "args": Array [
        "testbot",
        "This server was created Fri Apr 24 2020 at 13:04:49 UTC",
      ],
      "command": "rpl_created",
      "commandType": "reply",
      "prefix": "origin.irc.xx-net.org",
      "rawCommand": "003",
      "server": "origin.irc.xx-net.org",
    },
  ],
  Array [
    "raw",
    Object {
      "args": Array [
        "testbot",
        "origin.irc.xx-net.org",
        "UnrealIRCd-5.0.4",
        "iowrsxzdHtIDZRqpWGTSB",
        "lvhopsmntikraqbeIHzMQNRTOVKDdGLPZSCcf",
      ],
      "command": "rpl_myinfo",
      "commandType": "reply",
      "prefix": "origin.irc.xx-net.org",
      "rawCommand": "004",
      "server": "origin.irc.xx-net.org",
    },
  ],
  Array [
    "raw",
    Object {
      "args": Array [
        "testbot",
        "AWAYLEN=307",
        "CASEMAPPING=ascii",
        "CHANLIMIT=#:30",
        "CHANMODES=beI,kLf,lH,psmntirzMQNRTOVKDdGPZSCc",
        "CHANNELLEN=32",
        "CHANTYPES=#",
        "DEAF=d",
        "ELIST=MNUCT",
        "EXCEPTS",
        "EXTBAN=~,ptmTSOcarnqjf",
        "HCN",
        "INVEX",
        "are supported by this server",
      ],
      "command": "rpl_isupport",
      "commandType": "reply",
      "prefix": "origin.irc.xx-net.org",
      "rawCommand": "005",
      "server": "origin.irc.xx-net.org",
    },
  ],
  Array [
    "raw",
    Object {
      "args": Array [
        "testbot",
        "KICKLEN=307",
        "KNOCK",
        "MAP",
        "MAXCHANNELS=30",
        "MAXLIST=b:60,e:60,I:60",
        "MAXNICKLEN=30",
        "MINNICKLEN=0",
        "MODES=12",
        "NAMESX",
        "NETWORK=xx-Net",
        "NICKLEN=30",
        "PREFIX=(qaohv)~&@%+",
        "are supported by this server",
      ],
      "command": "rpl_isupport",
      "commandType": "reply",
      "prefix": "origin.irc.xx-net.org",
      "rawCommand": "005",
      "server": "origin.irc.xx-net.org",
    },
  ],
  Array [
    "raw",
    Object {
      "args": Array [
        "testbot",
        "QUITLEN=307",
        "SAFELIST",
        "SILENCE=15",
        "STATUSMSG=~&@%+",
        "TARGMAX=DCCALLOW:,ISON:,JOIN:,KICK:4,KILL:,LIST:,NAMES:1,NOTICE:1,PART:,PRIVMSG:4,SAJOIN:,SAPART:,USERHOST:,USERIP:,WATCH:,WHOIS:1,WHOWAS:1",
        "TOPICLEN=360",
        "UHNAMES",
        "USERIP",
        "WALLCHOPS",
        "WATCH=128",
        "WATCHOPTS=A",
        "WHOX",
        "are supported by this server",
      ],
      "command": "rpl_isupport",
      "commandType": "reply",
      "prefix": "origin.irc.corrupt-net.org",
      "rawCommand": "005",
      "server": "origin.irc.corrupt-net.org",
    },
  ],
  Array [
    "raw",
    Object {
      "args": Array [
        "testbot",
        "xx-xx.nothing.sanic.net",
        "is now your displayed host",
      ],
      "command": "396",
      "commandType": "normal",
      "prefix": "origin.irc.xx-net.org",
      "rawCommand": "396",
      "server": "origin.irc.xx-net.org",
    },
  ],
  Array [
    "raw",
    Object {
      "args": Array [
        "testbot",
        "*** You are connected to origin.irc.xx-net.org with TLSv1.3-TLS_CHACHA20_POLY1305_SHA256",
      ],
      "command": "NOTICE",
      "commandType": "normal",
      "prefix": "origin.irc.xx-net.org",
      "rawCommand": "NOTICE",
      "server": "origin.irc.xx-net.org",
    },
  ],
  Array [
    "notice",
    undefined,
    "testbot",
    "*** You are connected to origin.irc.xx-net.org with TLSv1.3-TLS_CHACHA20_POLY1305_SHA256",
    Object {
      "args": Array [
        "testbot",
        "*** You are connected to origin.irc.xx-net.org with TLSv1.3-TLS_CHACHA20_POLY1305_SHA256",
      ],
      "command": "NOTICE",
      "commandType": "normal",
      "prefix": "origin.irc.xx-net.org",
      "rawCommand": "NOTICE",
      "server": "origin.irc.xx-net.org",
    },
  ],
  Array [
    "raw",
    Object {
      "args": Array [
        "testbot",
        "There are 1 users and 530 invisible on 6 servers",
      ],
      "command": "rpl_luserclient",
      "commandType": "reply",
      "prefix": "origin.irc.xx-net.org",
      "rawCommand": "251",
      "server": "origin.irc.xx-net.org",
    },
  ],
  Array [
    "raw",
    Object {
      "args": Array [
        "testbot",
        "40",
        "operator(s) online",
      ],
      "command": "rpl_luserop",
      "commandType": "reply",
      "prefix": "origin.irc.xx-net.org",
      "rawCommand": "252",
      "server": "origin.irc.xx-net.org",
    },
  ],
  Array [
    "raw",
    Object {
      "args": Array [
        "testbot",
        "1",
        "unknown connection(s)",
      ],
      "command": "rpl_luserunknown",
      "commandType": "reply",
      "prefix": "origin.irc.xx-net.org",
      "rawCommand": "253",
      "server": "origin.irc.xx-net.org",
    },
  ],
  Array [
    "raw",
    Object {
      "args": Array [
        "testbot",
        "87",
        "channels formed",
      ],
      "command": "rpl_luserchannels",
      "commandType": "reply",
      "prefix": "origin.irc.xx-net.org",
      "rawCommand": "254",
      "server": "origin.irc.xx-net.org",
    },
  ],
  Array [
    "raw",
    Object {
      "args": Array [
        "testbot",
        "I have 170 clients and 1 servers",
      ],
      "command": "rpl_luserme",
      "commandType": "reply",
      "prefix": "origin.irc.xx-net.org",
      "rawCommand": "255",
      "server": "origin.irc.xx-net.org",
    },
  ],
  Array [
    "raw",
    Object {
      "args": Array [
        "testbot",
        "170",
        "411",
        "Current local users 170, max 411",
      ],
      "command": "rpl_localusers",
      "commandType": "reply",
      "prefix": "origin.irc.xx-net.org",
      "rawCommand": "265",
      "server": "origin.irc.xx-net.org",
    },
  ],
  Array [
    "raw",
    Object {
      "args": Array [
        "testbot",
        "531",
        "756",
        "Current global users 531, max 756",
      ],
      "command": "rpl_globalusers",
      "commandType": "reply",
      "prefix": "origin.irc.xx-net.org",
      "rawCommand": "266",
      "server": "origin.irc.xx-net.org",
    },
  ],
  Array [
    "raw",
    Object {
      "args": Array [
        "testbot",
        "MOTD File is missing",
      ],
      "command": "err_nomotd",
      "commandType": "error",
      "prefix": "origin.irc.xx-net.org",
      "rawCommand": "422",
      "server": "origin.irc.xx-net.org",
    },
  ],
  Array [
    "motd",
    "undefinedMOTD File is missing
",
  ],
  Array [
    "raw",
    Object {
      "args": Array [
        "testbot",
        "+iwxz",
      ],
      "command": "MODE",
      "commandType": "normal",
      "host": undefined,
      "nick": "testbot",
      "prefix": "testbot",
      "rawCommand": "MODE",
      "user": undefined,
    },
  ],
  Array [
    "raw",
    Object {
      "args": Array [
        "testbot",
        "This nickname is registered and protected.  If it is your",
      ],
      "command": "NOTICE",
      "commandType": "normal",
      "host": "nickserv.services.irc.xx-net.org",
      "nick": "NickServ",
      "prefix": "NickServ!nickserv@nickserv.services.irc.xx-net.org",
      "rawCommand": "NOTICE",
      "user": "nickserv",
    },
  ],
  Array [
    "notice",
    "NickServ",
    "testbot",
    "This nickname is registered and protected.  If it is your",
    Object {
      "args": Array [
        "testbot",
        "This nickname is registered and protected.  If it is your",
      ],
      "command": "NOTICE",
      "commandType": "normal",
      "host": "nickserv.services.irc.xx-net.org",
      "nick": "NickServ",
      "prefix": "NickServ!nickserv@nickserv.services.irc.xx-net.org",
      "rawCommand": "NOTICE",
      "user": "nickserv",
    },
  ],
  Array [
    "raw",
    Object {
      "args": Array [
        "testbot",
        "nick, type /msg NickServ IDENTIFY password.  Otherwise,",
      ],
      "command": "NOTICE",
      "commandType": "normal",
      "host": "nickserv.services.irc.xx-net.org",
      "nick": "NickServ",
      "prefix": "NickServ!nickserv@nickserv.services.irc.xx-net.org",
      "rawCommand": "NOTICE",
      "user": "nickserv",
    },
  ],
  Array [
    "notice",
    "NickServ",
    "testbot",
    "nick, type /msg NickServ IDENTIFY password.  Otherwise,",
    Object {
      "args": Array [
        "testbot",
        "nick, type /msg NickServ IDENTIFY password.  Otherwise,",
      ],
      "command": "NOTICE",
      "commandType": "normal",
      "host": "nickserv.services.irc.xx-net.org",
      "nick": "NickServ",
      "prefix": "NickServ!nickserv@nickserv.services.irc.xx-net.org",
      "rawCommand": "NOTICE",
      "user": "nickserv",
    },
  ],
  Array [
    "raw",
    Object {
      "args": Array [
        "testbot",
        "please choose a different nick.",
      ],
      "command": "NOTICE",
      "commandType": "normal",
      "host": "nickserv.services.irc.xx-net.org",
      "nick": "NickServ",
      "prefix": "NickServ!nickserv@nickserv.services.irc.xx-net.org",
      "rawCommand": "NOTICE",
      "user": "nickserv",
    },
  ],
  Array [
    "notice",
    "NickServ",
    "testbot",
    "please choose a different nick.",
    Object {
      "args": Array [
        "testbot",
        "please choose a different nick.",
      ],
      "command": "NOTICE",
      "commandType": "normal",
      "host": "nickserv.services.irc.xx-net.org",
      "nick": "NickServ",
      "prefix": "NickServ!nickserv@nickserv.services.irc.xx-net.org",
      "rawCommand": "NOTICE",
      "user": "nickserv",
    },
  ],
  Array [
    "raw",
    Object {
      "args": Array [
        "testbot",
        "If you do not change within 1 minute, I will change your nick.",
      ],
      "command": "NOTICE",
      "commandType": "normal",
      "host": "nickserv.services.irc.xx-net.org",
      "nick": "NickServ",
      "prefix": "NickServ!nickserv@nickserv.services.irc.xx-net.org",
      "rawCommand": "NOTICE",
      "user": "nickserv",
    },
  ],
  Array [
    "notice",
    "NickServ",
    "testbot",
    "If you do not change within 1 minute, I will change your nick.",
    Object {
      "args": Array [
        "testbot",
        "If you do not change within 1 minute, I will change your nick.",
      ],
      "command": "NOTICE",
      "commandType": "normal",
      "host": "nickserv.services.irc.xx-net.org",
      "nick": "NickServ",
      "prefix": "NickServ!nickserv@nickserv.services.irc.xx-net.org",
      "rawCommand": "NOTICE",
      "user": "nickserv",
    },
  ],
]
`;
